### DISCUSSION TITLE
trapezoid shape infeasible region (constrained optimization)

### DISCUSSION THREAD

**Discussion by minsu77** [meta-pytorch/botorch#1769] [Q&A]
I am trying to perform constrained Bayesian optimization using Botorch.
There is an inequality constraint like Case 1 in the attached file. In fact, an inequality constraint like Case 2 can be expressed using "inequality constraint" among the acquisition function commands.
However, when it is difficult to express with a linear inequality constraint, such as case 1, I am not sure how to express it.

I think it is an inequality nonlinear constraint. Do you have any examples?

** Code snippet to reproduce **
```python
    candidates, acq_value = optimize_acqf(
        acq_function=acq_func,
        bounds=standard_bounds,
        q=BATCH_SIZE,
        num_restarts=NUM_RESTARTS,
        raw_samples=RAW_SAMPLES,  
        options={"batch_limit": 5, "maxiter": 200},
        inequality_constraints=ineq_constraints, 
        sequential=True,
    )
    # observe new values 
    new_x =  unnormalize(candidates.detach(), bounds=standard_bounds)
    return new_x

ineq_constraints = [
            (torch.tensor([1, 2]), torch.tensor([-2.252, -1.]), torch.tensor([-2.752]))
        ]
```
![image](https://user-images.githubusercontent.com/46668422/228286005-504784a1-c4af-4ec9-bec3-348406bb638e.png)

## System information

**Please complete the following information:**
- 
- 
- 
-

**Comment by Balandat** [comment]
The main challenge with your constraint is that it's non-convex, so it can't be expressed an intersection of half spaces (or linear inequality constraints). The "good" news is that the acquisition function optimization is already non-convex so this doesn't necessarily make things a lot harder from an optimization perspective. You can use the (experimental) support for non-linear inequality constraints for this: https://github.com/pytorch/botorch/blob/main/botorch/optim/optimize.py#L438-L445

Basically the callable that you need to provide will need be positive when outside the infeasible region - so you could use e.g. the distance to the infeasible set.
